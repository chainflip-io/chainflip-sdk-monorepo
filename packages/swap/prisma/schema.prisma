// This is your Prisma schema file,
// learn more about it in the docs: https://pris.ly/d/prisma-schema

// `pnpm prisma format` after changes :)

generator client {
  provider        = "prisma-client-js"
  // https://github.com/prisma/prisma/issues/2584#issuecomment-646414549
  // https://github.com/prisma/prisma/issues/6603#issuecomment-849979849
  output          = "../node_modules/.prisma/client"
  previewFeatures = ["multiSchema"]
}

datasource db {
  provider = "postgresql"
  url      = env("DATABASE_URL")
  schemas  = ["public", "private"]
}

enum Chain {
  Polkadot
  Ethereum
  Bitcoin

  @@schema("public")
}

enum Asset {
  FLIP
  USDC
  DOT
  ETH
  BTC

  @@schema("public")
}

model SwapDepositChannel {
  id                    BigInt   @id @default(autoincrement())
  uuid                  String   @unique @default(uuid())
  srcAsset              Asset
  depositAddress        String
  expectedDepositAmount Decimal  @db.Decimal(30, 0)
  destAsset             Asset
  destAddress           String
  expiryBlock           Int
  issuedBlock           Int
  createdAt             DateTime @default(now())
  swaps                 Swap[]

  @@index([depositAddress])
  @@schema("public")
}

model Swap {
  id                        BigInt              @id @default(autoincrement())
  /// the swap id on the state chain
  nativeId                  BigInt              @unique
  srcAsset                  Asset
  depositAmount             Decimal              @db.Decimal(30, 0)
  intermediateAmount        Decimal?             @db.Decimal(30, 0)
  egressAmount              Decimal?             @db.Decimal(30, 0)
  destAsset                 Asset
  destAddress               String
  depositReceivedAt         DateTime
  depositReceivedBlockIndex String
  swapExecutedAt            DateTime?
  swapExecutedBlockIndex    String?
  egressCompletedAt         DateTime?
  egressCompletedBlockIndex String?
  swapDepositChannel        SwapDepositChannel? @relation(fields: [swapDepositChannelId], references: [id])
  swapDepositChannelId      BigInt?
  txHash                    String?             @unique
  egress                    Egress?
  createdAt                 DateTime            @default(now())
  updatedAt                 DateTime            @default(now()) @updatedAt

  @@schema("public")
}

model Egress {
  id          BigInt     @id @default(autoincrement())
  /// the egress id on the state chain
  nativeId    BigInt
  /// the swap id on the state chain
  chain       Chain
  amount      Decimal    @db.Decimal(30, 0)
  timestamp   DateTime
  swap        Swap?      @relation(fields: [swapId], references: [id])
  swapId      BigInt?    @unique
  broadcast   Broadcast? @relation(fields: [broadcastId], references: [id])
  broadcastId BigInt?
  createdAt   DateTime   @default(now())

  @@unique([nativeId, chain])
  @@schema("public")
}

model Broadcast {
  id       BigInt   @id @default(autoincrement())
  /// the broadcast id on the state chain
  nativeId BigInt
  chain    Chain
  egresses Egress[]

  @@unique([nativeId, chain])
  @@schema("public")
}

model State {
  id     Int @id @default(autoincrement())
  height Int @default(0)

  @@schema("private")
}

model MarketMaker {
  id        Int    @id @default(autoincrement())
  name      String @unique
  publicKey String

  @@schema("private")
}

enum ThirdPartyProtocol {
  lifi
  squid

  @@schema("public")
}

model ThirdPartySwap {
  id            BigInt             @id @default(autoincrement())
  uuid          String             @unique
  protocol      ThirdPartyProtocol
  routeResponse Json
  txHash        String
  txLink        String
  createdAt     DateTime           @default(now())

  @@schema("public")
}
