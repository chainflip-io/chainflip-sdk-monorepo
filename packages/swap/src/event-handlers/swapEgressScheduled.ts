import { z } from 'zod';
import { Asset, assetChains } from '@/shared/enums';
import { bigintMin } from '@/shared/functions';
import { chainflipChain, unsignedInteger } from '@/shared/parsers';
import { Environment, getEnvironment } from '@/shared/rpc';
import { readAssetValue } from '@/shared/rpc/utils';
import { CacheMap } from '@/swap/utils/dataStructures';
import { estimateIngressEgressFeeAssetAmount } from '@/swap/utils/fees';
import env from '../config/env';
import type { EventHandlerArgs } from '.';

const eventArgs = z.object({
  swapId: unsignedInteger,
  egressId: z.tuple([
    z.object({ __kind: chainflipChain }).transform(({ __kind }) => __kind),
    unsignedInteger,
  ]),
});

export type SwapEgressScheduledEvent = z.input<typeof eventArgs>;

const environmentByBlockHashCache = new CacheMap<
  string,
  Promise<Environment | null>
>(60_000);

const methodNotFoundRegExp = /Exported method .+ is not found/;
const rpcConfig = { rpcUrl: env.RPC_NODE_HTTP_URL };

const getCachedEnvironmentAtBlock = async (
  blockHash: string,
): Promise<Environment | null> => {
  const cached = environmentByBlockHashCache.get(blockHash);
  if (cached) return cached;

  const environment = getEnvironment(rpcConfig, blockHash).catch((e: Error) => {
    const cause = e.cause as { message: string } | undefined;

    if (methodNotFoundRegExp.test(cause?.message ?? '')) return null;

    environmentByBlockHashCache.delete(blockHash);
    throw e;
  });

  environmentByBlockHashCache.set(blockHash, environment);

  return environment;
};

const assetAmountByNativeAmountAndBlockHashCache = new CacheMap<
  string,
  Promise<bigint>
>(60_000);

const getCachedAssetAmountAtBlock = async (
  asset: Asset,
  nativeAmount: bigint,
  blockHash: string,
): Promise<bigint> => {
  const cacheKey = `${asset}-${nativeAmount.toString()}-${blockHash}`;
  const cached = assetAmountByNativeAmountAndBlockHashCache.get(cacheKey);
  if (cached) return cached;

  const rate = estimateIngressEgressFeeAssetAmount(
    nativeAmount,
    asset,
    blockHash,
  ).catch((e: Error) => {
    assetAmountByNativeAmountAndBlockHashCache.delete(blockHash);
    throw e;
  });

  assetAmountByNativeAmountAndBlockHashCache.set(cacheKey, rate);

  return rate;
};

const getEgressFeeAtBlock = async (
  blockHash: string,
  asset: Asset,
): Promise<bigint> => {
  const environment = await getCachedEnvironmentAtBlock(blockHash);
  if (!environment) return 0n;

  const nativeFee = readAssetValue(environment.ingressEgress.egressFees, {
    asset,
    chain: assetChains[asset],
  });

  return getCachedAssetAmountAtBlock(asset, nativeFee, blockHash);
};

/**
 * this event is emitted in order to correlate the egress id from a network
 * deposit/egress pallet to a swap id
 */
export default async function swapEgressScheduled({
  prisma,
  event,
  block,
}: EventHandlerArgs): Promise<void> {
  const {
    swapId,
    egressId: [chain, nativeId],
  } = eventArgs.parse(event.args);

  const swap = await prisma.swap.findUniqueOrThrow({
    where: { nativeId: swapId },
  });

  // TODO: use an accurate source for determining the egress fee once the protocol provides one
  const egressFee = bigintMin(
    await getEgressFeeAtBlock(block.hash, swap.destAsset),
    BigInt(swap.swapOutputAmount?.toFixed() ?? 0),
  );

  await Promise.all([
    prisma.egress.update({
      where: { nativeId_chain: { chain, nativeId } },
      data: {
        amount: swap.swapOutputAmount?.sub(egressFee.toString()),
      },
    }),
    prisma.swap.update({
      where: { nativeId: swapId },
      data: {
        egress: { connect: { nativeId_chain: { chain, nativeId } } },
        fees: {
          create: {
            type: 'EGRESS',
            asset: swap.destAsset,
            amount: egressFee.toString(),
          },
        },
      },
    }),
  ]);
}
