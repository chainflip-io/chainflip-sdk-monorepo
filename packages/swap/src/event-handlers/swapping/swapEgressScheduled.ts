import { swappingSwapEgressScheduled as schema180 } from '@chainflip/processor/180/swapping/swapEgressScheduled';
import { swappingSwapEgressScheduled as schema190 } from '@chainflip/processor/190/swapping/swapEgressScheduled';
import { chainConstants } from '@chainflip/utils/chainflip';
import { z } from 'zod';
import type { EventHandlerArgs } from '..';

const eventArgs = z.union([schema190, schema180]);

export type SwapEgressScheduledArgs = z.input<typeof eventArgs>;

/**
 * this event is emitted in order to correlate the egress id from a network
 * deposit/egress pallet to a swap id
 */
export default async function swapEgressScheduled({
  prisma,
  event,
  block,
}: EventHandlerArgs): Promise<void> {
  const {
    swapRequestId,
    egressId: [chain, nativeId],
    egressFee,
    amount,
    asset,
  } = eventArgs.parse(event.args);

  const request = await prisma.swapRequest.findUniqueOrThrow({
    where: { nativeId: swapRequestId },
  });

  const egressFeeAsset =
    egressFee[1] ?? (request.requestType === 'LEGACY_CCM' ? chainConstants[chain].gasAsset : asset);

  await prisma.swapRequest.update({
    where: { nativeId: swapRequestId },
    data: {
      egress: {
        create: {
          nativeId,
          chain,
          amount: amount.toString(),
          scheduledAt: new Date(block.timestamp),
          scheduledBlockIndex: `${block.height}-${event.indexInBlock}`,
        },
      },
      fees: {
        create: {
          type: 'EGRESS',
          asset: egressFeeAsset,
          amount: egressFee[0].toString(),
        },
      },
    },
  });
}
