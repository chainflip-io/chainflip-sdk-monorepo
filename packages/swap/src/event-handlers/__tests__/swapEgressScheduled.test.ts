import prisma from '../../client';
import swapEgressScheduled from '../swapEgressScheduled';
import {
  DOT_ADDRESS,
  createDepositChannel,
  swapEgressScheduledMock,
} from './utils';

const {
  eventContext: { event },
  block,
} = swapEgressScheduledMock;

describe(swapEgressScheduled, () => {
  beforeEach(async () => {
    await prisma.$queryRaw`TRUNCATE TABLE "SwapDepositChannel", "Swap" CASCADE`;
  });

  it('updates an existing swap with the scheduled timestamp', async () => {
    const { swapId } = event.args;

    await prisma.egress.create({
      data: {
        network: event.args.egressId[0].__kind,
        nativeId: BigInt(event.args.egressId[1]),
        amount: '1234567890',
        timestamp: new Date(block.timestamp),
      },
    });

    // store a new swap intent to initiate a new swap
    const swapDepositChannel = await createDepositChannel({
      swaps: {
        create: {
          nativeId: BigInt(swapId),
          depositAmount: '10000000000',
          depositReceivedAt: new Date(block.timestamp - 12000),
          depositReceivedBlockIndex: `${block.height - 100}-${
            event.indexInBlock
          }`,
          swapExecutedAt: new Date(block.timestamp - 6000),
          swapExecutedBlockIndex: `${block.height}-${event.indexInBlock}`,
          srcAsset: 'ETH',
          destAsset: 'DOT',
          destAddress: DOT_ADDRESS,
        },
      },
    });

    await prisma.$transaction((tx) =>
      swapEgressScheduled({
        block: block as any,
        event: event as any,
        prisma: tx,
      }),
    );

    const swap = await prisma.swap.findFirstOrThrow({
      where: { swapDepositChannelId: swapDepositChannel.id },
      include: {
        egress: {
          select: {
            amount: true,
            timestamp: true,
            network: true,
          },
        },
      },
    });

    expect(swap).toMatchSnapshot({
      id: expect.any(BigInt),
      createdAt: expect.any(Date),
      updatedAt: expect.any(Date),
      swapDepositChannelId: expect.any(BigInt),
    });
  });
});
