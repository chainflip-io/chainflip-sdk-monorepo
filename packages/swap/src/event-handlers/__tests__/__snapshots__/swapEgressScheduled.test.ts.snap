// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`swapEgressScheduled > adds an egress and fee to a swap request 1`] = `
{
  "ccmDepositReceivedBlockIndex": null,
  "ccmGasBudget": null,
  "ccmMessage": null,
  "chunkIntervalBlocks": null,
  "completedAt": null,
  "completedBlockIndex": null,
  "depositAmount": "10000000000",
  "depositBoostedAt": null,
  "depositBoostedBlockIndex": null,
  "depositFinalisedAt": 2022-12-06T14:31:33.000Z,
  "depositFinalisedBlockIndex": "20-0",
  "depositTransactionRef": null,
  "destAddress": "1yMmfLti1k3huRQM2c47WugwonQMqTvQ2GUFxnU7Pcs7xPo",
  "destAsset": "Dot",
  "effectiveBoostFeeBps": null,
  "egress": {
    "amount": "4945198948008612506",
    "broadcastId": null,
    "chain": "Ethereum",
    "createdAt": Any<Date>,
    "id": Any<BigInt>,
    "nativeId": 1n,
    "scheduledAt": 2022-12-06T14:31:45.000Z,
    "scheduledBlockIndex": "120-0",
    "updatedAt": Any<Date>,
  },
  "egressId": Any<BigInt>,
  "fees": [
    {
      "amount": "167509500781711",
      "asset": "Eth",
      "type": "EGRESS",
    },
  ],
  "id": Any<BigInt>,
  "maxBoostFeeBps": 0,
  "nativeId": 9876545n,
  "numberOfChunks": null,
  "originType": "VAULT",
  "prewitnessedDepositId": null,
  "refundEgressId": null,
  "requestType": "LEGACY_SWAP",
  "srcAddress": null,
  "srcAsset": "Eth",
  "swapDepositChannelId": null,
  "swapInputAmount": "10000000000",
  "swapOutputAmount": null,
  "swapRequestedAt": 2022-12-06T14:31:33.000Z,
  "swapRequestedBlockIndex": "92-398",
  "totalBrokerCommissionBps": 0,
}
`;

exports[`swapEgressScheduled > adds an egress and fee to a swap request for 160 schema 1`] = `
{
  "ccmDepositReceivedBlockIndex": null,
  "ccmGasBudget": null,
  "ccmMessage": null,
  "chunkIntervalBlocks": null,
  "completedAt": null,
  "completedBlockIndex": null,
  "depositAmount": "10000000000",
  "depositBoostedAt": null,
  "depositBoostedBlockIndex": null,
  "depositFinalisedAt": 2022-12-06T14:31:33.000Z,
  "depositFinalisedBlockIndex": "20-0",
  "depositTransactionRef": null,
  "destAddress": "1yMmfLti1k3huRQM2c47WugwonQMqTvQ2GUFxnU7Pcs7xPo",
  "destAsset": "Dot",
  "effectiveBoostFeeBps": null,
  "egress": {
    "amount": "10000000000",
    "broadcastId": null,
    "chain": "Ethereum",
    "createdAt": Any<Date>,
    "id": Any<BigInt>,
    "nativeId": 1n,
    "scheduledAt": 2022-12-06T14:31:45.000Z,
    "scheduledBlockIndex": "120-0",
    "updatedAt": Any<Date>,
  },
  "egressId": Any<BigInt>,
  "fees": [
    {
      "amount": "1000000000",
      "asset": "Eth",
      "type": "EGRESS",
    },
  ],
  "id": Any<BigInt>,
  "maxBoostFeeBps": 0,
  "nativeId": 9876545n,
  "numberOfChunks": null,
  "originType": "VAULT",
  "prewitnessedDepositId": null,
  "refundEgressId": null,
  "requestType": "LEGACY_SWAP",
  "srcAddress": null,
  "srcAsset": "Eth",
  "swapDepositChannelId": null,
  "swapInputAmount": "10000000000",
  "swapOutputAmount": null,
  "swapRequestedAt": 2022-12-06T14:31:33.000Z,
  "swapRequestedBlockIndex": "92-398",
  "totalBrokerCommissionBps": 0,
}
`;

exports[`swapEgressScheduled > uses gas asset for egress fee when CCM for 160 schema 1`] = `
{
  "ccmDepositReceivedBlockIndex": null,
  "ccmGasBudget": null,
  "ccmMessage": null,
  "chunkIntervalBlocks": null,
  "completedAt": null,
  "completedBlockIndex": null,
  "depositAmount": "1000000",
  "depositBoostedAt": null,
  "depositBoostedBlockIndex": null,
  "depositFinalisedAt": 2022-12-06T14:31:33.000Z,
  "depositFinalisedBlockIndex": "20-0",
  "depositTransactionRef": null,
  "destAddress": "0x6Aa69332B63bB5b1d7Ca5355387EDd5624e181F2",
  "destAsset": "ArbUsdc",
  "effectiveBoostFeeBps": null,
  "egress": {
    "amount": "10000000000",
    "broadcastId": null,
    "chain": "Arbitrum",
    "createdAt": Any<Date>,
    "id": Any<BigInt>,
    "nativeId": 1n,
    "scheduledAt": 2022-12-06T14:31:45.000Z,
    "scheduledBlockIndex": "120-0",
    "updatedAt": Any<Date>,
  },
  "egressId": Any<BigInt>,
  "fees": [
    {
      "amount": "1000000000",
      "asset": "ArbEth",
      "type": "EGRESS",
    },
  ],
  "id": Any<BigInt>,
  "maxBoostFeeBps": 0,
  "nativeId": 9876545n,
  "numberOfChunks": null,
  "originType": "VAULT",
  "prewitnessedDepositId": null,
  "refundEgressId": null,
  "requestType": "CCM",
  "srcAddress": null,
  "srcAsset": "ArbUsdc",
  "swapDepositChannelId": null,
  "swapInputAmount": "1000000",
  "swapOutputAmount": null,
  "swapRequestedAt": 2022-12-06T14:31:33.000Z,
  "swapRequestedBlockIndex": "92-398",
  "totalBrokerCommissionBps": 0,
}
`;
